### ISSUE TITLE
Deployment Error "Serverless: Cannot read property 'replace' of undefined"??

### ISSUE THREAD

**Issue by contractorwolf** [serverless/serverless#1153]
I am trying to deploy a simple GET request using version 0.5.5 and I keep having these weird deployment issues.  This is a test endpoint that just returns the default but it wont deploy now, hits this mysterious error:
Serverless:   GET - ping: Cannot read property 'replace' of undefined  

here is the terminal output from attempting to deploy:
`
Serverless: Select the assets you wish to deploy:
    ping
      function - ping
      endpoint - ping - GET
    - - - - -
  Deploy
    Cancel
  serverless:lib.node_modules.serverless.lib.utils.cli Closing select listener +3s

Serverless: Deploying the specified functions in "dev" to the following regions: us-east-1  
  serverless:lib.node_modules.serverless.lib.Runtime "dev - us-east-1 - ping": Copying in dist dir /Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta/_tmp/ping@1463891945509 +10ms
  serverless:lib.node_modules.serverless.lib.Runtime "dev - us-east-1 - ping": Excluding - Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta +4ms
  serverless:lib.node_modules.serverless.lib.Runtime "dev - us-east-1 - ping": Excluding - Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/admin.env +0ms
Serverless: \   serverless:lib.node_modules.serverless.lib.actions.FunctionDeploy codePackageLambda is done +265ms
  serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda compressing the function bundle +3ms
Serverless: /   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Compressed file created - /Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta/_tmp/ping_dev_us-east-1.zip +682ms
Serverless: -   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Updating Lambda configuration... +339ms
Serverless: /   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Updating Lambda function... +702ms
Serverless: -   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Updating Lambda Alias for version - 4 +3s
Serverless: /   serverless:lib.node_modules.serverless.lib.actions.FunctionDeploy code deployment is done +199ms
  serverless:lib.node_modules.serverless.lib.actions.FunctionDeploy Removing `_tmp` folder +1ms
Serverless: ------------------------  
Serverless: Successfully deployed the following functions in "dev" to the following regions:   
Serverless: us-east-1 ------------------------  
Serverless:   ping (sls-content-admin-ping): arn:aws:lambda:us-east-1:237169730438:function:sls-content-admin-ping:dev  

Serverless: Deploying endpoints in "dev" to the following regions: us-east-1  
Serverless: -   serverless:lib.node_modules.serverless.lib.ProviderAws "dev - us-east-1": found existing REST API on AWS API Gateway with name: sls-content-admin +377ms
  serverless:lib.node_modules.serverless.lib.utils.index Writing file: /Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta/variables/s-variables-dev-useast1.json... +4ms
Serverless: Failed to deploy endpoints in "dev" to the following regions:  
Serverless: us-east-1 ------------------------  
Serverless:   GET - ping: Cannot read property 'replace' of undefined  
  serverless:lib.node_modules.serverless.lib.actions.EndpointDeploy TypeError: Cannot read property 'replace' of undefined
    at ServerlessProviderAws.getAccountId (/usr/local/lib/node_modules/serverless/lib/ProviderAws.js:352:11)
    at Builder._validateAndPrepare (/usr/local/lib/node_modules/serverless/lib/actions/EndpointBuildApiGateway.js:123:46)
    at Builder.build (/usr/local/lib/node_modules/serverless/lib/actions/EndpointBuildApiGateway.js:69:20)
    at EndpointBuildApiGateway.endpointBuildApiGateway (/usr/local/lib/node_modules/serverless/lib/actions/EndpointBuildApiGateway.js:53:22)
    at processImmediate [as _immediateCallback] (timers.js:371:17)
From previous event:
    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:211:25
    at Array.reduce (native)
    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:210:26)
    at Object.actions.(anonymous function) [as endpointBuildApiGateway] (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)
    at /usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:172:36
From previous event:
    at /usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:146:35
    at /usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:181:20
    at iterate (/usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:262:13)
    at Object.async.forEachOfSeries.async.eachOfSeries (/usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:281:9)
    at Object.async.forEachSeries.async.eachSeries (/usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:214:22)
    at /usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:144:19
From previous event:
    at EndpointDeployApiGateway.<anonymous> (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:141:18)
From previous event:
    at EndpointDeployApiGateway._processRegionDeployment (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:137:10)
    at EndpointDeployApiGateway.<anonymous> (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:94:24)
From previous event:
    at EndpointDeployApiGateway._processDeployment (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:91:10)
From previous event:
    at EndpointDeployApiGateway.endpointDeployApiGateway (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:55:10)
From previous event:
    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:211:25
    at Array.reduce (native)
    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:210:26)
    at Object.actions.(anonymous function) [as endpointDeployApiGateway] (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)
    at EndpointDeploy._processDeployment (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeploy.js:212:24)
From previous event:
    at EndpointDeploy.endpointDeploy (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeploy.js:101:12)
From previous event:
    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:211:25
    at Array.reduce (native)
    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:210:26)
    at Object.actions.(anonymous function) [as endpointDeploy] (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)
    at /usr/local/lib/node_modules/serverless/lib/actions/DashDeploy.js:251:28
From previous event:
    at DashDeploy._deploy (/usr/local/lib/node_modules/serverless/lib/actions/DashDeploy.js:246:10)
From previous event:
    at DashDeploy.dashDeploy (/usr/local/lib/node_modules/serverless/lib/actions/DashDeploy.js:126:10)
From previous event:
    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:186:31
    at Array.reduce (native)
    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:185:32
From previous event:
    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:183:12)
    at Serverless.actions.(anonymous function) (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)
    at Serverless.command (/usr/local/lib/node_modules/serverless/lib/Serverless.js:398:38)
    at /usr/local/lib/node_modules/serverless/bin/serverless:19:16
    at processImmediate [as _immediateCallback] (timers.js:371:17)
From previous event:
    at Object.<anonymous> (/usr/local/lib/node_modules/serverless/bin/serverless:18:4)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:475:10)
    at startup (node.js:117:18)
    at node.js:951:3 +130ms
Serverless:   
Serverless: Run this again with --debug to get more error information...  
serverless: Select the assets you wish to deploy:    ping      function - ping      endpoint - ping - GET    - - - - -  > Deploy    Cancel  serverless:lib.node_modules.serverless.lib.utils.cli Closing select listener +3sServerless: Deploying the specified functions in "dev" to the following regions: us-east-1    serverless:lib.node_modules.serverless.lib.Runtime "dev - us-east-1 - ping": Copying in dist dir /Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta/_tmp/ping@1463891945509 +10ms  serverless:lib.node_modules.serverless.lib.Runtime "dev - us-east-1 - ping": Excluding - Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta +4ms  serverless:lib.node_modules.serverless.lib.Runtime "dev - us-east-1 - ping": Excluding - Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/admin.env +0msServerless: \   serverless:lib.node_modules.serverless.lib.actions.FunctionDeploy codePackageLambda is done +265ms  serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda compressing the function bundle +3msServerless: /   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Compressed file created - /Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta/_tmp/ping_dev_us-east-1.zip +682msServerless: -   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Updating Lambda configuration... +339msServerless: /   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Updating Lambda function... +702msServerless: -   serverless:lib.node_modules.serverless.lib.actions.CodeDeployLambda "dev - us-east-1 - sls-content-admin-ping": Updating Lambda Alias for version - 4 +3sServerless: /   serverless:lib.node_modules.serverless.lib.actions.FunctionDeploy code deployment is done +199ms  serverless:lib.node_modules.serverless.lib.actions.FunctionDeploy Removing `_tmp` folder +1msServerless: ------------------------  Serverless: Successfully deployed the following functions in "dev" to the following regions:   Serverless: us-east-1 ------------------------  Serverless:   ping (sls-content-admin-ping): arn:aws:lambda:us-east-1:237169730438:function:sls-content-admin-ping:dev  Serverless: Deploying endpoints in "dev" to the following regions: us-east-1  Serverless: -   serverless:lib.node_modules.serverless.lib.ProviderAws "dev - us-east-1": found existing REST API on AWS API Gateway with name: sls-content-admin +377ms  serverless:lib.node_modules.serverless.lib.utils.index Writing file: /Users/jameswolf/Documents/nodejs/serverless/sls-content-admin/_meta/variables/s-variables-dev-useast1.json... +4msServerless: Failed to deploy endpoints in "dev" to the following regions:  Serverless: us-east-1 ------------------------  Serverless:   GET - ping: Cannot read property 'replace' of undefined    serverless:lib.node_modules.serverless.lib.actions.EndpointDeploy TypeError: Cannot read property 'replace' of undefined    at ServerlessProviderAws.getAccountId (/usr/local/lib/node_modules/serverless/lib/ProviderAws.js:352:11)    at Builder._validateAndPrepare (/usr/local/lib/node_modules/serverless/lib/actions/EndpointBuildApiGateway.js:123:46)    at Builder.build (/usr/local/lib/node_modules/serverless/lib/actions/EndpointBuildApiGateway.js:69:20)    at EndpointBuildApiGateway.endpointBuildApiGateway (/usr/local/lib/node_modules/serverless/lib/actions/EndpointBuildApiGateway.js:53:22)    at processImmediate [as _immediateCallback] (timers.js:371:17)From previous event:    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:211:25    at Array.reduce (native)    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:210:26)    at Object.actions.(anonymous function) [as endpointBuildApiGateway] (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)    at /usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:172:36From previous event:    at /usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:146:35    at /usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:181:20    at iterate (/usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:262:13)    at Object.async.forEachOfSeries.async.eachOfSeries (/usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:281:9)    at Object.async.forEachSeries.async.eachSeries (/usr/local/lib/node_modules/serverless/node_modules/async/lib/async.js:214:22)    at /usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:144:19From previous event:    at EndpointDeployApiGateway.<anonymous> (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:141:18)From previous event:    at EndpointDeployApiGateway._processRegionDeployment (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:137:10)    at EndpointDeployApiGateway.<anonymous> (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:94:24)From previous event:    at EndpointDeployApiGateway._processDeployment (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:91:10)From previous event:    at EndpointDeployApiGateway.endpointDeployApiGateway (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeployApiGateway.js:55:10)From previous event:    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:211:25    at Array.reduce (native)    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:210:26)    at Object.actions.(anonymous function) [as endpointDeployApiGateway] (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)    at EndpointDeploy._processDeployment (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeploy.js:212:24)From previous event:    at EndpointDeploy.endpointDeploy (/usr/local/lib/node_modules/serverless/lib/actions/EndpointDeploy.js:101:12)From previous event:    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:211:25    at Array.reduce (native)    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:210:26)    at Object.actions.(anonymous function) [as endpointDeploy] (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)    at /usr/local/lib/node_modules/serverless/lib/actions/DashDeploy.js:251:28From previous event:    at DashDeploy._deploy (/usr/local/lib/node_modules/serverless/lib/actions/DashDeploy.js:246:10)From previous event:    at DashDeploy.dashDeploy (/usr/local/lib/node_modules/serverless/lib/actions/DashDeploy.js:126:10)From previous event:    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:186:31    at Array.reduce (native)    at /usr/local/lib/node_modules/serverless/lib/Serverless.js:185:32From previous event:    at Serverless._execute (/usr/local/lib/node_modules/serverless/lib/Serverless.js:183:12)    at Serverless.actions.(anonymous function) (/usr/local/lib/node_modules/serverless/lib/Serverless.js:429:20)    at Serverless.command (/usr/local/lib/node_modules/serverless/lib/Serverless.js:398:38)    at /usr/local/lib/node_modules/serverless/bin/serverless:19:16    at processImmediate [as _immediateCallback] (timers.js:371:17)From previous event:    at Object.<anonymous> (/usr/local/lib/node_modules/serverless/bin/serverless:18:4)    at Module._compile (module.js:434:26)    at Object.Module._extensions..js (module.js:452:10)    at Module.load (module.js:355:32)    at Function.Module._load (module.js:310:12)    at Function.Module.runMain (module.js:475:10)    at startup (node.js:117:18)    at node.js:951:3 +130msServerless:   Serverless: Run this again with --debug to get more error information...

`

**Comment by bradgreens** [comment]
I'm experiencing something similar to this. I have a different callstack, however the error is traced to the exact same point in the stack, which is `ServerlessProviderAws.getAccountId `.

I had my AWS CLI configured before I ever installed serverless, and my hunch is something inside serverless doesn't like how my account is setup. I'm not sure what to do yet. Might nuke my project and rebuild.

Any advice from the team regarding `ServerlessProviderAws.getAccountId` being undefined for event deployments? Oddly enough I can deploy my functions just fine. And luckily our other devs environment "just works" and we're skating past the issue unfortunately.

**Comment by contractorwolf** [comment]
Have you tried deleting your _meta folder and then use "sls project init" command? That will establish the correct AWS permissions. Sometimes I create a new profile with my AWS details.


Get Outlook for Android






On Fri, May 27, 2016 at 11:53 AM -0700, "Brad Greenwald" <notifications@github.com> wrote:





I'm experiencing something similar to this. I have a different callstack, however the error is traced to the exact same point in the stack, which is `ServerlessProviderAws.getAccountId `.

I had my AWS CLI configured before I ever installed serverless, and my hunch is something inside serverless doesn't like how my account is setup. I'm not sure what to do yet. Might nuke my project and rebuild.

Any advice from the team regarding `ServerlessProviderAws.getAccountId` being undefined for event deployments? Oddly enough I can deploy my functions just fine. And luckily our other devs environment "just works" and we're skating past the issue unfortunately.

---

**Comment by mrpudn** [comment]
I am also getting this error.

We can't create IAM roles in our environment, but we do have an existing IAM role that we can use, so I've set that in my `s-function.json` here:
```json
{
  ...
  "customRole": "$${iamRoleArnLambda}",
  ...
}
```

I have `iamRoleArnLambda` defined in my project's `s-templates.json`.

We have a theory that our IAM role might not have the permissions to perform these actions (which doesn't sound quite right to me), but we're still looking into that.

**Comment by mrpudn** [comment]
Ok, it looks like I was able to deploy everything successfully by changing my `_meta/variables/s-variables-dev-useast1.json` file to include my IAM role arm:
```json
{
  "region": "us-east-1",
  "iamRoleArnLambda": "<my-iam-role-arn>"
  }
```

Now, that's a pretty kludgy workaround, so if anyone has a better idea of how to get around this, I'd love to know! :)

**Comment by kendaop** [comment]
We were running into this exact same issue. I was trying to create another endpoint for a resource that was already created, but was getting this error. I tried deleting my _meta folder and creating the project files again, with no luck. For some reason, the `iamRoleArnLambda` property wasn't getting created in my `_meta/variables/s-variables-dev-useast1.json` file. I had to copy the values from my co-worker (who initially set up the project), to get it to work.

**Comment by gebrits** [comment]
@kendaop: seeing this issue as well and think it's a bug, see below.

afaik the `iarnRoleArnLambda` attrib should be created by doing `sls resources deploy`

Having said that, the following flow doesn't: 

After having done a `sls init` on a different dev pc, I did the above `sls resources deploy`. This gave me the message: 

```
Serverless: Deploying resources to stage "dev" in region "us-east-1" via Cloudformation (~3 minutes)...  
Serverless: No resource updates are to be performed.
```

In other words, SLS determines that cloudformation config on the server is up to date and nothing needs to change. Therefore it shortcircuits. However, and I'm guessing here, it shortcircuits too much: it should still create/overwrite the `iarnRoleArnLambda`

Anyone can confirm?

**Comment by gebrits** [comment]
Had a quick dig in sourcecode: 

The fact that CloudFormation is up-to-date is regarded as an exception. 

I.e.: "No resource updates are to be performed" is thrown here.
https://github.com/serverless/serverless/blob/078a0281183cf1c376b2e95994326d8ba1ce2b5c/lib/actions/ResourcesDeploy.js#L305

This results, in the `.then` of the promise not being executed. It is here that the `iarnRoleArnLambda` is set. 

In short , this is a bug and is easily solved by not treating 'cloudformation up to dateness' as an exception but rather a no-op. I'll do a pullrequest when I'm able

**Comment by Fandekasp** [comment]
Thanks @gebrits, was having the same problem, and running `sls resource deploy` did the trick (although the behavior is surprising, by running deploy I'd expect the server to be updated from the local version, not the other way around)

**Comment by flomotlik** [comment]
Hi, thank you for reaching out to us! At this point we've stopped working on 0.5 of Serverless to fully focus on V1.0 that we're currently working on. Therefore we're not taking any changes for 0.5 and do not plan any future releases at this point. We only have a limited set of resources and just need to make sure we're putting everything we have into our next release (and can provide great long term support and backwards compatibility from then on). Hope you understand.

We would love to get your feedback on V1 of the framework and I'm happy to walk you through any questions to make sure its all working great for you.

**Comment by karthikknidhi** [comment]
Make sure all the environment variables on your YAML file are exported

**Comment by Rakshith658** [comment]
add `configValidationMode: error` to serverless.yml file and `run serverless deploy
if 
`Error:
Configuration error at 'functions.hello.events.1.http': value 'ANY' does not satisfy pattern /^(?:\*|(GET|POST|PUT|PATCH|OPTIONS|HEAD|DELETE|ANY) (\/\S*))$/i` 
then check the spacing properly in serverless.yml in events
`events:
      - http: ANY /{proxy+}
      - http: ANY /`
